File: problems/链表/23.合并K个升序链表.py
#!/usr/bin/env python3
"""
23. 合并K个升序链表 (Merge K Sorted Lists)
题目: 给你一个链表数组，每个链表都已经按升序排列。请你将所有链表合并成一个升序链表并返回。
"""

import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def merge_k_lists(lists) -> ListNode:
    heap = []
    for i, node in enumerate(lists):
        if node:
            heapq.heappush(heap, (node.val, i, node))
    
    dummy = ListNode(0)
    cur = dummy
    while heap:
        val, i, node = heapq.heappop(heap)
        cur.next = node
        cur = cur.next
        if node.next:
            heapq.heappush(heap, (node.next.val, i, node.next))
    return dummy.next

def list_to_linkedlist(lst):
    dummy = ListNode(0)
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

if __name__ == "__main__":
    # 测试用例1: [[1,4,5],[1,3,4],[2,6]] 合并为 [1,1,2,3,4,4,5,6]
    lists = [
        list_to_linkedlist([1, 4, 5]),
        list_to_linkedlist([1, 3, 4]),
        list_to_linkedlist([2, 6])
    ]
    result = merge_k_lists(lists)
    print("测试1 - 合并 [[1,4,5],[1,3,4],[2,6]]:")
    print(f"  输出: {linkedlist_to_list(result)}")  # [1, 1, 2, 3, 4, 4, 5, 6]

    # 测试用例2: [[], [1]] 合并为 [1]
    lists = [None, list_to_linkedlist([1])]
    result = merge_k_lists(lists)
    print("\n测试2 - 合并 [[], [1]]:")
    print(f"  输出: {linkedlist_to_list(result)}")  # [1]

    # 测试用例3: [[1,2,3],[4,5,6]] 合并为 [1,2,3,4,5,6]
    lists = [
        list_to_linkedlist([1, 2, 3]),
        list_to_linkedlist([4, 5, 6])
    ]
    result = merge_k_lists(lists)
    print("\n测试3 - 合并 [[1,2,3],[4,5,6]]:")
    print(f"  输出: {linkedlist_to_list(result)}")  # [1, 2, 3, 4, 5, 6]

File: problems/链表/23.合并K个升序链表.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* listToLinkedList(vector<int> lst) {
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    for (int val : lst) {
        cur->next = new ListNode(val);
        cur = cur->next;
    }
    return dummy->next;
}

vector<int> linkedListToList(ListNode* node) {
    vector<int> result;
    while (node) {
        result.push_back(node->val);
        node = node->next;
    }
    return result;
}

struct cmp {
    bool operator()(ListNode* a, ListNode* b) {
        return a->val > b->val;
    }
};

ListNode* mergeKLists(vector<ListNode*> lists) {
    priority_queue<ListNode*, vector<ListNode*>, cmp> pq;
    for (auto node : lists) {
        if (node) pq.push(node);
    }
    
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    while (!pq.empty()) {
        ListNode* node = pq.top();
        pq.pop();
        cur->next = node;
        cur = cur->next;
        if (node->next) {
            pq.push(node->next);
        }
    }
    return dummy->next;
}

void printList(ListNode* node) {
    cout << "[";
    while (node) {
        cout << node->val;
        if (node->next) cout << ",";
        node = node->next;
    }
    cout << "]";
}

int main() {
    // 测试用例1: [[1,4,5],[1,3,4],[2,6]] 合并为 [1,1,2,3,4,4,5,6]
    vector<ListNode*> lists1 = {
        listToLinkedList({1, 4, 5}),
        listToLinkedList({1, 3, 4}),
        listToLinkedList({2, 6})
    };
    ListNode* result = mergeKLists(lists1);
    cout << "测试1 - 合并 [[1,4,5],[1,3,4],[2,6]]:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    // 测试用例2: [[], [1]] 合并为 [1]
    vector<ListNode*> lists2 = {nullptr, listToLinkedList({1})};
    result = mergeKLists(lists2);
    cout << "\n测试2 - 合并 [[], [1]]:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    // 测试用例3: [[1,2,3],[4,5,6]] 合并为 [1,2,3,4,5,6]
    vector<ListNode*> lists3 = {
        listToLinkedList({1, 2, 3}),
        listToLinkedList({4, 5, 6})
    };
    result = mergeKLists(lists3);
    cout << "\n测试3 - 合并 [[1,2,3],[4,5,6]]:" << endl;
    cout << "  输出: ";
    printList(result);
    cout << endl;

    return 0;
}
